test/topology_custom/test_table_desc_read_barrier.py

```diff
@@ -0,0 +1,61 @@
+#
+# Copyright (C) 2024-present ScyllaDB
+#
+# SPDX-License-Identifier: AGPL-3.0-or-later
+#
+
+import logging
+import time
+import pytest
+
+from test.pylib.manager_client import ManagerClient
+from test.pylib.rest_client import inject_error
+from test.pylib.util import wait_for_cql_and_get_hosts
+from test.topology.conftest import skip_mode
+
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.mark.asyncio
+@skip_mode('release', 'error injections are not supported in release mode')
+async def test_table_desc_read_barrier(request, manager: ManagerClient) -> None:
+    """
+    Regression test for #19213.
+
+    Test that the schema is always updated after read barrier.
+
+    Arrange:
+        - start 2 nodes
+        - create a table
+        - inject error (sleep) on the first node
+        - disable the schema agreement wait
+    Act:
+        - alter the table (change schema) on the second node
+        - trigger the read barrier
+    Assert:
+        - the schema retrieved on both hosts matches
+    """
+    servers = await manager.servers_add(2)
+
+    logging.info("Waiting until driver connects to every server")
+    cql = manager.get_cql()
+    # this is needed so that the hosts will be matching the servers
+    hosts = [(await wait_for_cql_and_get_hosts(cql, [s], time.time() + 60))[0] for s in servers]
+
+    logger.info("Creating keyspace and table")
+    await cql.run_async("create keyspace ks with replication = "
+                        "{'class': 'NetworkTopologyStrategy', 'replication_factor': 1}")
+    await cql.run_async("create table ks.t (pk int primary key)")
+
+    logger.info("Disabling the schema agreement wait")
+    assert hasattr(cql.cluster, "max_schema_agreement_wait")
+    cql.cluster.max_schema_agreement_wait = 0
+
+    async with inject_error(manager.api, servers[0].ip_addr, 'group0_state_machine::delay_apply'):
+        logger.info("Altering table")
+        await cql.run_async("alter table ks.t add s1 int", host=hosts[1])
+        # TODO: trigger the read barrier once available
```

Comment:
Does this test pass at this stage, with this TODO?
If it doesn't, it breaks bisectability.
Maybe you should just add the test, in its complete form, as the second patch and not the first?

Reply:
The test passes (so it should not break bisectability) - the expectation is that the schema is different without the read barrier (which the test checks here), after adding the barrier the assertion is switched for equality

(initially it was just switching the condition to equality after adding the barrier, however later I actually kept the first check for unequality to make sure the initial assumption holds, and then triggering the read barrier and check for equality of the schemas)

But it might indeed make sense to switch the order (or even squash both to one commit).

Reply:
changed to comparison for equality (as is expected later) and marked the test `xfail` in this commit (that makes it "pass as failed"), the `xfail` is then removed in the impl commit

in particular, the test has been made first to make sure the issue is reproduced first (to ensure that the read barrier addition actually fixes the problem, kind of in a TDD sense)